src/vw/src/main/scala/VowpalWabbitBase.scala

```diff
@@ -0,0 +1,393 @@
+// Copyright (C) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in project root for information.
+
+package com.microsoft.ml.spark
+
+import java.util.UUID
+
+import org.apache.spark.{ClusterUtil, TaskContext}
+import org.apache.spark.ml.linalg.{DenseVector, SparseVector}
+import org.apache.spark.ml.param._
+import org.apache.spark.ml.util.DefaultParamsWritable
+import org.apache.spark.sql.functions.{col, struct, udf}
+import org.apache.spark.sql.types.StructType
+import org.apache.spark.sql.{DataFrame, Dataset, Encoders, Row}
+import org.apache.spark.internal._
+import org.vowpalwabbit.spark.prediction.ScalarPrediction
+import org.vowpalwabbit.spark.{ClusterSpanningTree, VowpalWabbitExample, VowpalWabbitMurmur, VowpalWabbitNative}
+
+import scala.math.min
+
+case class NamespaceInfo (hash: Int, featureGroup: Char, colIdx: Int)
+
+/**
+  * Base implementation of VowpalWabbit learners.
+  *
+  * @note parameters that regularly are swept through are exposed as proper parameters.
+  */
+@InternalWrapper
+trait VowpalWabbitBase extends Wrappable
+  with DefaultParamsWritable
+  with HasWeightCol
+  with Logging
+{
+  // can we switch to https://docs.scala-lang.org/overviews/macros/paradise.html ?
+  val args = new Param[String](this, "args", "VW command line arguments passed")
+  setDefault(args -> "")
+
+  def getArgs: String = $(args)
+  def setArgs(value: String): this.type = set(args, value)
+
+  // VW support multiple input columns which are mapped to namespaces.
+  val additionalFeatures = new StringArrayParam(this, "additionalFeatures", "Additional feature columns")
+  setDefault(additionalFeatures -> Array())
+  def getAdditionalFeatures: Array[String] = $(additionalFeatures)
+  def setAdditionalFeatures(value: Array[String]): this.type = set(additionalFeatures, value)
+
+  // to support Grid search we need to replicate the parameters here...
+  val numPasses = new IntParam(this, "numPasses", "Number of passes over the data")
+  setDefault(numPasses -> 1)
+  def getNumPasses: Int = $(numPasses)
+  def setNumPasses(value: Int): this.type = set(numPasses, value)
+
+  val learningRate = new DoubleParam(this, "learningRate", "Learning rate")
+  def getLearningRate: Double = $(learningRate)
+  def setLearningRate(value: Double): this.type = set(learningRate, value)
+
+  val powerT = new DoubleParam(this, "powerT", "t power value")
+  def getPowerT: Double = $(powerT)
+  def setPowerT(value: Double): this.type = set(powerT, value)
+
+  val l1 = new DoubleParam(this, "l1", "l_1 lambda")
```

Comment:
you don't need any defaults set for these?

Reply:
Your catching all those nice details… so I don’t want to replicate the default values from the C++ codebase as they might change.
Thus I’m only conditionally passing the values if provide.

Added a comment.